Moving NETStandard2.0 netfx support into SDK

ericstj opened the pull request:
Replaces https://github.com/dotnet/sdk/pull/1273

/cc @dsplaisted @livarcocc

ericstj committed:
Add task for determining if NETStandard is referenced

ericstj committed:
Add a separate build of tasks for NETFramework

ericstj committed:
Add NETStandard.Library.NETFramework content

ericstj committed:
Condition inclusion of NS2 netfx facades.

ericstj committed:
Support override of targets file location

dsplaisted committed:
Merge upstream changes

ericstj committed:
Use TargetFrameworkVersion comparison, check for inbox

Future-proof this target by turning off the functionality once netstandard.dll is inbox.

Improve conditions by doing version comparison instead of string equality checks.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GetDependsOnNETStandard.netstandard.cs:
```diff
@@ -0,0 +1,43 @@
+                            AssemblyReference reference = reader.GetAssemblyReference(referenceHandle);
+                            var referenceName = reader.GetString(reference.Name);
+
+                            if (result = referenceName.Equals(NetStandardAssemblyName, StringComparison.Ordinal))
```
Nit: I think  it would clearer to return true here and false outside the loop. The local + assignment is hard to follow IMHO and unnecessary.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GetDependsOnNETStandard.cs:
```diff
@@ -0,0 +1,73 @@
+                {
+                    try
+                    {
+                        if (DependsOnNETStandard = GetFileDependsOnNETStandard(referenceSourcePath))
```
Similar nit here. Factor out a method that loops over references and returns true/false and then assign result of that to DependsOnNETStandard.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GetDependsOnNETStandard.net46.cs:
```diff
@@ -0,0 +1,163 @@
+            }
+            else
+            {
+                // on Windows use CLR's unmanaged metadata API.
```
We should comment on why we're not using S.R.M here. I think I know the answer, but it should be noted here.

ericstj committed:
Rename to Microsoft.NET.Build.Extensions

Rename the folder to Microsoft.NET.Build.Extensions since we intend for it
to apply for more than just desktop in the future.

Rename the task assembly and ensure it cross-compiles for core as well.

ericstj committed:
Clarify GetDependsOnNETStandard task

Refactor GetDependsOnNETStandard a bit and add some comments.

ericstj committed:
Change hook to after MS.Common.targets

We'll need to extend more than just NETFramework projects so move
the hook to after MS.Common.targets.

Create a central targets file that conditionally imports NETFramework targets
and rename them as well.

ericstj committed:
Add conflict resolution for non-SDK projects

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GetDependsOnNETStandard.cs:
```diff
@@ -28,6 +28,13 @@ public partial class GetDependsOnNETStandard : TaskBase
         {
+            DependsOnNETStandard = AnyReferenceDependsOnNETStandard();
+
+            return;
```
nit: remove unncessary `return;`

ericstj replied:
The single assignment felt lonely.

nguerrera replied:
Lonely code is good code ;)

ericstj committed:
Clarify comments in Build.Extensions targets

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/GetDependsOnNETStandard.cs:
```diff
@@ -0,0 +1,80 @@
+    /// <summary>
+    /// Determines the assembly version to use for a given semantic version.
+    /// </summary>
+    public partial class GetDependsOnNETStandard : TaskBase
```
I could move this entirely into the Build.Extensions dll since it is only used there now that I changed the tasks to be imported independently.

ericstj replied:
@dsplaisted and I discussed and decided to move it.  I'll need to create a separate test project to do that.

ericstj committed:
Add Microsoft.NET.Build.Extensions.Tasks to sln

ericstj commented:
This should be code-complete now.  Looking into testing.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/msbuildExtensions/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.targets:
```diff
@@ -0,0 +1,13 @@
+    <MicrosoftNETBuildExtensionsTasksAssembly Condition="'$(MicrosoftNETBuildExtensionsTasksAssembly)' == ''">$(MSBuildThisFileDirectory)\tools\net46\Microsoft.NET.Build.Extensions.Tasks.dll</MicrosoftNETBuildExtensionsTasksAssembly>
+  </PropertyGroup>
+
+  <Import Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'" Project="Microsoft.NET.Build.Extensions.NETFramework.targets"/>
```
Microsoft.Common.CurrentVersion.targets [uses the following condition](https://github.com/Microsoft/msbuild/blob/5fcb6c489a2039a008d7820139db1dff33bb7a03/src/Tasks/Microsoft.Common.CurrentVersion.targets#L5780) to import Microsoft.NETFramework.targets:

```
Condition="('$(TargetFrameworkIdentifier)' == ''  or '$(TargetFrameworkIdentifier)' == '.NETFramework') and ('$(TargetRuntime)' == 'Managed')"
```

Should we use the same condition here in order to work with projects that don't explicitly set the TargetFrameworkIdentifier (which I think may be most projects)?

ericstj replied:
TargetFrameworkIdentifier will be set by common targets by the time we're imported.

eerhardt commented:
I'm getting the same full framework test failures in my PR.  @livarcocc @nguerrera @dsplaisted - is the branch currently broken?

nguerrera commented:
I think https://github.com/dotnet/sdk/pull/1281 should have fixed it

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/msbuildExtensions/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.ConflictResolution.targets:
```diff
@@ -0,0 +1,55 @@
+  </Choose>
+
+  <Target Name="PackagingToolsInsufficientSDKVersion" Condition="'$(PackagingToolsInsufficientSDKVersion)' != 'false'" AfterTargets="$(PackagingToolsInsufficientSDKVersionAfter)">
+    <Error Text="The version of Microsoft.NET.Sdk used by this project is insufficient to support this package.  Please ensure you are using the .NET Core SDK 2.0 or higher, which is included in Visual Studio 2017 Update 15.3.  This error may be suppressed by setting PackagingToolsInsufficientSDKVersion=false." />
```
VS 15.3 won't necessarily ship with the .NET Core SDK in-box, it may only ship with a 1.1 version of the SDK.  So if you haven't installed the 2.0 SDK, I think as is you'll always get this error message when you try to build an sdk-style project using full MSBuild.

On the other hand, if you haven't installed the 2.0 SDK but are targeting .NET 4.6.1, we should either make it work, or provide a good error.  We need to figure out when & how to raise the error, and the error text needs to be changed (since the 2.0 SDK won't necessarily be in VS 15.3) and put into a .resx for localization (which we do with the NETSdkError task).

ericstj replied:
We can't make it work on 1.1 SDK without bringing back deps file re-writing.  We discussed this and decided to raise the error in the NETStandard injection targets, only when netstandard is referenced and running on old SDK.

ericstj replied:
I've done this.  I had to create a different task name so that it wouldn't collide with the SDK one.

ericstj commented:
@dotnet-bot test Windows_NT_FullFramework Debug

ericstj commented:
@dotnet-bot test Windows_NT_FullFramework Release

ericstj committed:
Merge 'release/2.0.0' into netstandard-netfx

ericstj committed:
Remove GetDependsOnNETStandard from main task assembly

This task is only used from the extensions assembly.

ericstj committed:
Error when using ns2.0 reference in netfx project on old SDK

It's possible that someone updates VS/nuget but doesn't update their SDK,
in which case we want to emit an error that they need to update the .NET SDK.

ericstj commented:
PTAL

dsplaisted committed:
Update error message when 2.0 version of SDK is not installed

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/msbuildExtensions/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets:
```diff
@@ -0,0 +1,102 @@
+    
+    <!-- if any reference depends on netstandard and it is not inbox, add references and implementation assemblies for netstandard2.0  -->
+    <ItemGroup Condition="'$(DependsOnNETStandard)' == 'true' AND '$(NETStandardInbox)' != 'true'">
+      <_NETStandardLibraryNETFrameworkReference Condition="'$(_TargetFrameworkVersionWithoutV)' &gt;= '4.7'"
```
I'd like to see an explanation of how this DLL selection works.  It looks like it doesn't have a way for a higher framework to remove a DLL, which we would have needed if we shipped something like System.ValueTuple in 4.7.

ericstj replied:
RAR will remove the DLL if it becomes inbox in the framework.  We don't remove DLLs within subsequent out-of-band releases.  This is existing logic that's just been reimplemented from the package.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/msbuildExtensions/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.ConflictResolution.targets:
```diff
@@ -0,0 +1,45 @@
+  <Target Name="HandlePackageFileConflicts" AfterTargets="$(HandlePackageFileConflictsAfter)" DependsOnTargets="$(HandlePackageFileConflictsDependsOn)">
+    <HandlePackageFileConflicts References="@(Reference)"
+                                ReferenceCopyLocalPaths="@(ReferenceCopyLocalPaths)"
+                                PlatformManifests="@(PackageConflictPlatformManifests)"
```
Do the `PlatformManifests` and `PreferredPackages` parameters apply here?  Don't those only apply when targeting .NET Core or .NET Standard, in which the SDK target would be used instead?

ericstj replied:
Its common behavior that should be exposed regardless of TFM.  You are correct in that we don't make use of it by default for .NETFramework so these will be empty.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/NETSdkError.cs:
```diff
@@ -2,16 +2,20 @@
     /// </summary>
-    public class NETSdkError : TaskBase
+    public class
+#if EXTENSIONS
```
Is there a reason this class needs a different name for each assembly?  If so, it would be nice to have a comment explaining why this is.

ericstj replied:
Yes, I mentioned that in the commit.  I can add a comment.

dsplaisted committed:
Add Extensions unit tests to solution and build

dsplaisted commented:
@MattGertz for approval

**Customer scenario**

Use a .NET Standard 2.0 library from a .NET Framework project.

**Bugs this fixes:** 

#1286

**Workarounds, if any**

n/a

**Risk**

Medium - This is hooking into the MSBuild reference pipeline, in different ways depending on the project type or how it references NuGet packages.

**Performance impact**

Low to Medium: This avoids doing work when possible, but has to inspect referenced DLLs to determine whether they depend on netstandard.dll

**Is this a regression from a previous update?**

No

**Root cause analysis:**

n/a - this is feature work

**How was the bug found?**

n/a - this is feature work

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/msbuildExtensions-ver/Microsoft.Common.targets/ImportAfter/Microsoft.NET.Build.Extensions.targets:
```diff
@@ -0,0 +1,7 @@
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <MicrosoftNETBuildExtensionsTargets Condition="'$(MicrosoftNETBuildExtensionsTargets)' == ''">$(MSBuildExtensionsPath)\Microsoft.NET.Build.Extensions\Microsoft.NET.Build.Extensions.targets</MicrosoftNETBuildExtensionsTargets>
```
I am concerned about the risk of adding targets to the build of all 15.3 .NET Framework projects (including classic csproj) at this stage of the 15.3 cycle.

* Have we adequately measured the performance impact given that this can impact all .NET framework projects, and hence a much larger set of customer projects.

* Do we have the approval that will be required to put this additional feature in-box in 15.3? I don't think it meets the bar that @MattGertz sent out recently without a special exception. 

I'm also unhappy with the complexity this adds to risk assessment of code changes in this repo moving forward. Previously, we could say that any change in the release/2.0.0 branch was scoped to the out-of-band .NET Core 2.0 SDK install, but now these targets are destined to go in-box, and so we'd have to audit each change to see the true impact and to know what level of approval we need. Furthermore, with the shared source files, it's not easy to see what changes are headed in-box and what changes are not. :(

Previously, insertions from CLI -> VS carried only 1.1 CLI and 2.0 resolver. This was simple to reason about. With this, we'd also have a chunk of the 2.0 SDK code going in to that payload. I want to be absolutely sure that the important decision makers that pushed to limit the churn from .NET Core 2.0 to 15.3 are fully aware of this before we move forward with it.

ericstj replied:
We have approval, folks specifically pushed for the design change to put this stuff inbox.

I share your concern for performance, see below.

As far as risk assessment, what is your recommendation?  I discussed various options with @dsplaisted and this is where we landed.

nguerrera replied:
The performance numbers are reassuring. Thank you for measuring and providing the details.

For risk assessment, what were the alternatives that were discussed?

ericstj replied:
I had discussed forking more of the code so that a minimal set was shared, which we could move to a common folder.  Also discussed building a separate package for this stuff.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Extensions.Tasks/Microsoft.NET.Build.Extensions.Tasks.csproj:
```diff
@@ -0,0 +1,75 @@
+    <EmbeddedResource Include="..\Microsoft.NET.Build.Tasks\Resources\Strings.*.resx">
+      <Link>Resources\%(FileName)%(Extension)</Link>
+      <DependentUpon>Strings.resx</DependentUpon>
+    </EmbeddedResource>
```
This going to conflict with my loc automation change. And it has alerted me to a case I had not considered for the update-xlf-on-build option (which we only turn on in IDE builds): if two projects share the same resx and build in parallel then the update can race.

ericstj replied:
Let's just disable it for the Extensions project and make sure it builds after the main project.

nguerrera replied:
OK, that should work. Arrange for UpdateXlfOnBuild to always be false for this project after merging my change. Slap a TreatAsLocal=UpdateXlfOnBuild so that build /p:UpdateXlfOnBuild=true still works on the command line. Do you know if ResXFileGenerator -> Designer.cs can have the same problem (I vaguely recall that it can). We should probably exclude that metadata too.

eerhardt commented on src/Tasks/Microsoft.NET.Build.Extensions.Tasks/GetDependsOnNETStandard.cs:
```diff
@@ -0,0 +1,80 @@
+namespace Microsoft.NET.Build.Tasks
+{
+    /// <summary>
+    /// Determines the assembly version to use for a given semantic version.
```
This comment seems like a copy-paste error.

ericstj replied:
Good catch, will fix.

nguerrera commented:
Regarding performance, it seems very unfortunate that there will be two passes to determine if (a) depends on System.Runtime and (b) depends on netstandard. If we're putting targets in-box anyway, it seems that we should be amending core targets to check for netstandard in the same pass as system.runtime. The duplicated target here could check if it's running downlevel from an msbuild with knowledge of the new core assembly name. 

I am also curious what the overhead of conflict resolution is for a .NET framework project that doesn't need it.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Extensions.Tasks/GetDependsOnNETStandard.netstandard.cs:
```diff
@@ -0,0 +1,42 @@
+                            AssemblyReference reference = reader.GetAssemblyReference(referenceHandle);
+                            var referenceName = reader.GetString(reference.Name);
+
+                            if (referenceName.Equals(NetStandardAssemblyName, StringComparison.Ordinal))
```
You can use reader.StringComparer.Equals to check this without allocating a string for each reference.

ericstj replied:
Ok, will look into that.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/msbuildExtensions/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.targets:
```diff
@@ -0,0 +1,17 @@
+    <MicrosoftNETBuildExtensionsTasksAssembly Condition="'$(MicrosoftNETBuildExtensionsTasksAssembly)' == ''">$(MSBuildThisFileDirectory)\tools\net46\Microsoft.NET.Build.Extensions.Tasks.dll</MicrosoftNETBuildExtensionsTasksAssembly>
+
+    <!-- Include conflict resolution targets for NETFramework and allow other frameworks to opt-in -->
+    <ResolveAssemblyConflicts Condition="'$(ResolveAssemblyConflicts)' == '' AND '$(TargetFrameworkIdentifier)' == '.NETFramework' AND '$(UsingMicrosoftNETSdk)' != 'true'">true</ResolveAssemblyConflicts>
```
By default, can this trigger only in the references-netstandard case to avoid adding overhead to all .NET framework projects?

ericstj replied:
/cc @terrajobst 
That's something to consider.  Previously we had said we wanted conflict resolution to be a characteristic of any build so that its easier to reason about.  Would it help if I shared some numbers that illustrate the overhead?

nguerrera replied:
Numbers would certainly help. Have it being part of all builds is goodness, but I want to understand the perf tradeoff.

nguerrera replied:
We're probably ok if the measurements you provided below included conflict resolution.

ericstj replied:
Those were measurements I made early on of just the task that was being added.  I went back and measured the targets in their current form, results are here: https://gist.github.com/ericstj/b56c6a430e79e6768f93551b66219d22

This is for a packages.config-based desktop project building in visual studio.  It's showing zero-impact to the hello-world case, and 16 ms cold / 0 ms warm to the top-10 packages (20 assemblies) case.  I've tried a few different variations of this but can't get numbers worse than this for these projects, so I'm fairly confident that we'll be able to keep the impact low.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/msbuildExtensions/Microsoft.NET.Build.Extensions/Microsoft.NET.Build.Extensions.NETFramework.targets:
```diff
@@ -0,0 +1,102 @@
+          AfterTargets="$(ImplicitlyExpandNETStandardFacadesAfter)">
+
+    <ItemGroup>
+      <_candidateNETStandardReferences Include="@(Reference);@(_ResolvedProjectReferencePaths)" />
```
nit: we use capital letters after _ elsehwere.

ericstj commented:
We've tuned for performance.  Impact to hello-world is near zero since it has no additional references.  Impact to a project referencing top 10 nuget packages (19 assemblies) was 9ms cold and 6ms warm for the new task.  That's one of the reasons why I ported the COR metadata code.  Without that we saw much larger times using metadata reader (66ms cold) due to load/JIT costs.

RAR is currently the thing responsible for detecting system.runtime.  At that point it is too late to add references.  We need to resolve conflicts between these additions and other package files, and they need to be passed as an input to RAR.  Certainly we can put both conflict resolution and the facade addition into RAR, but that seems like a much larger change to make at this point.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Extensions.Tasks/GetDependsOnNETStandard.net46.cs:
```diff
@@ -0,0 +1,169 @@
+{
+    public partial class GetDependsOnNETStandard
+    {
+        private static Guid s_importerGuid = new Guid("7DAC8207-D3AE-4c75-9B67-92801A497D44");
```
readonly

nguerrera commented on src/Tasks/Microsoft.NET.Build.Extensions.Tasks/GetDependsOnNETStandard.net46.cs:
```diff
@@ -0,0 +1,169 @@
+                // on Windows use CLR's unmanaged metadata API.
+                // Create the metadata dispenser and open scope on the source file.
+                IMetaDataDispenser metadataDispenser = (IMetaDataDispenser)new CorMetaDataDispenser();
+                IMetaDataAssemblyImport assemblyImport = (IMetaDataAssemblyImport)metadataDispenser.OpenScope(filePath, 0, ref s_importerGuid);
```
nit: `var` since type is obvious from cast.

ericstj replied:
Ok, this code was copied so I tried to make minimal changes.  I went through and sprinkled `var` where it was appropriate.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Extensions.Tasks/Microsoft.NET.Build.Extensions.Tasks.csproj:
```diff
@@ -0,0 +1,75 @@
+    <TargetFrameworks>netcoreapp1.0;net46</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netcoreapp1.0</TargetFrameworks>
+    <GenerateDependencyFile>false</GenerateDependencyFile>
+    <OutputPath>$(OutputPath)Sdks\Microsoft.NET.Sdk\msbuildExtensions\Microsoft.NET.Build.Extensions\tools</OutputPath>
```
Upon further thought, we should make this its own package. Doing it this way means the CLI build will have to do filtering on what it puts in the Microsoft.NET.Sdk folder. These are logically separate components that live together because they share code. They should be easy to trace through as distinct things in the CLI setup code.

dsplaisted replied:
I've been thinking it should be all in one package so there's a single insertion from the SDK to the CLI and we don't have to change how  that process works.  I'm coming around to thinking it would be better to have two separate packages though, as I look at how the layout would need to be filtered by the CLI.

nguerrera replied:
Please make them separate.

nguerrera replied:
There isn't really anything difficult about using more than 1 package to insert in to CLI. The build definition for SDK pushes bin\Release\Packages\*.nupkg already. And you'd just need to restore and use the extra package in the CLI build. The two packages would be used in different places in the CLI build so there's really no win I can see to keeping them together in 1 package.

ericstj replied:
I'll look into doing this.

ericstj replied:
I moved it into a separate package.  PTAL.  Also let me know if you'd like me to split the shared code into some common folder.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Extensions.Tasks.UnitTests/GivenAGetDependsOnNETStandardTask.cs:
```diff
@@ -0,0 +1,152 @@
+                File.Delete(lockedFile);
+            }
+        }
+    }
```
There should also be end-to-end tests that actually launch msbuild for .NET Framework (classic and sdk) -> .NET Standard 2 refs. These can be modeled after other tests in Microsoft.NET.Build.Tests.

dsplaisted replied:
We have #1287 with on overview of the test scenarios we want.  I'd like to merge these changes as soon as they're ready so that we can work on the insertion into the CLI and VS in parallel while Eric works on those tests.

nguerrera replied:
Ok

ericstj committed:
Respond to feedback / fix non-sdk projects

ericstj committed:
Merge remote-tracking branch 'upstream/netstandard-netfx' into netstandard-netfx

ericstj committed:
Build MS.NET.Build.Extensions as its own package

Rather than putting all of this in the same package as the SDK, build a separate package.

ericstj committed:
Merge branch 'release/2.0.0' into netstandard-netfx

# Conflicts:
#	build/DependencyVersions.props
#	src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/Microsoft.NET.Build.Tasks.UnitTests.csproj

ericstj committed:
Update XLF files with added strings

Result of manually running /t:UpdateXlf on Microsoft.NET.Build.Tasks.

ericstj committed:
Fix build issues after merging release/2.0.0

ericstj committed:
Don't use SLN to sequence build

It appears that a SLN that defines project dependencies doesn't build correctly on Unix
so use a ProjectReference to ensure that Microsoft.NET.Build.Extensions.Tasks and
Microsoft.NET.Build.Tasks don't build concurrently (and both try to update Xlf files).

ericstj committed:
Move shared source to common folder

ericstj commented:
@dotnet-bot test Windows_NT Release

ericstj commented:
@dotnet-bot test Windows_NT_FullFramework Release

nguerrera commented on build/Signing/SignToolConfig.json:
```diff
@@ -31,7 +31,35 @@
-                "Sdks/Microsoft.NET.Sdk/tools/net46/zh-Hant/Microsoft.NET.Build.Tasks.resources.dll"
+                "Sdks/Microsoft.NET.Sdk/tools/net46/zh-Hant/Microsoft.NET.Build.Tasks.resources.dll",
+                "Sdks/Microsoft.NET.Build.Extensions/msbuildExtensions/Microsoft.NET.Build.Extensions/tools/netcoreapp1.0/Microsoft.NET.Build.Tasks.dll",
+                "Sdks/Microsoft.NET.Build.Extensions/msbuildExtensions/Microsoft.NET.Build.Extensions/tools/netcoreapp1.0/cs/Microsoft.NET.Build.Tasks.resources.dll",
```
Good catch on signing, but it looks like the assembly names still needs to be edited.

ericstj replied:
Doh.

ericstj committed:
Fix signing and resource assembly build for Extensions
